File: java/solitaire/game/src/main/java/org/silnith/game/search/GameTreeSearcher.java
package org.silnith.game.search;

import java.io.PrintStream;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.atomic.AtomicLong;

import org.silnith.game.Game;
import org.silnith.game.GameState;
import org.silnith.game.move.Move;
import org.silnith.game.move.MoveFilter;
import org.silnith.util.LinkedNode;

/**
 * Abstract base class for game tree search algorithms.
 * This provides a common implementation of the logic to expand a single node
 * in the game tree.
 * 
 * @param <M> the move type for the game
 * @param <B> the board type for the game
 */
public abstract class GameTreeSearcher<M extends Move<B>, B> implements AutoCloseable, Callable<Collection<List<GameState<M, B>>>> {

    private final Game<M, B> game;
    private final Collection<? extends MoveFilter<M, B>> gameFilters;
    private final AtomicLong gameStatesExamined;
    private final AtomicLong boardsGenerated;
    private final AtomicLong movesPrunedTotal;
    private final Map<Object, AtomicLong> movesPruned;

    /**
     * Initializes a game tree search run for the provided game.
     * 
     * @param game the game
     */
    protected GameTreeSearcher(final Game<M, B> game) {
        super();
        if (game == null) {
            throw new IllegalArgumentException("Game cannot be null.");
        }
        this.game = game;
        this.gameFilters = this.game.getFilters();
        this.gameStatesExamined = new AtomicLong();
        this.boardsGenerated = new AtomicLong();
        this.movesPrunedTotal = new AtomicLong();
        final Map<Object, AtomicLong> tempMap = new HashMap<>();
        for (final MoveFilter<M, B> filter : gameFilters) {
            final Object statisticsKey = filter.getStatisticsKey();
            tempMap.put(statisticsKey, new AtomicLong());
        }
        this.movesPruned = Collections.unmodifiableMap(tempMap);
    }

    /**
     * Returns the current size of the queue of nodes to be searched.
     * 
     * @return the current size of the queue
     */
    protected abstract int getQueueSize();

    /**
     * Returns the current number of winning game states found.
     * 
     * @return the current number of wins found
     */
    protected abstract int getWinCount();

    /**
     * Prints statistics on the game tree search to the provided
     * print stream.
     * 
     * @param out the print stream to print statistics to
     */
    public void printStatistics(final PrintStream out) {
        out.printf(Locale.US,
                "Nodes examined: %,d\n"
                + "Boards generated: %,d\n"
                + "Moves pruned: %,d\n"
                + "Queue size: %,d\n"
                + "Wins: %,d\n",
                gameStatesExamined.get(),
                boardsGenerated.get(),
                movesPrunedTotal.get(),
                getQueueSize(),
                getWinCount());
        for (final MoveFilter<M, B> filter : gameFilters) {
            final Object statisticsKey = filter.getStatisticsKey();
            out.printf(Locale.US,
                    "Moves pruned by filter %s: %,d\n",
                    statisticsKey,
                    movesPruned.get(statisticsKey).get());
        }
        out.flush();
    }
    
    /**
     * Returns the total number of game states examined.
     * 
     * @return the total number of game states examined.
     */
    public long getNumberOfGameStatesExamined() {
        return gameStatesExamined.get();
    }
    
    /**
     * Returns the total number of boards generated.
     * 
     * @return the total number of boards generated
     */
    public long getBoardsGenerated() {
        return boardsGenerated.get();
    }

    /**
     * Queues a game tree node for searching.
     * 
     * @param node the game tree node to search
     */
    protected abstract void queueNode(LinkedNode<GameState<M, B>> node);

    /**
     * Captures a sequence of game states that culminates in a win.
     * 
     * @param node the game state history that results in a win
     */
    protected abstract void addWin(List<GameState<M, B>> node);

    /**
     * Examines a single node in the game tree.  This enumerates all possible
     * moves, applies the move filters, and examines unfiltered moves to see if
     * they are a winning game state.  Winning game states are passed to
     * {@link #addWin(List)}.  All other unfiltered game states are passed to
     * {@link #queueNode(LinkedNode)}.
     * 
     * <p>This also captures statistics on the number of nodes examined, boards
     * generated, and moves filtered.  These statistics can be printed using
     * {@link #printStatistics(PrintStream)}.
     * 
     * @param node the node containing the game state to examine
     */
    protected void examineNode(final LinkedNode<GameState<M, B>> node) {
        gameStatesExamined.getAndIncrement();
        final GameState<M, B> gameState = node.getValue();
        final B board = gameState.getBoard();
        final Collection<M> moves = game.findAllMoves(node);
        movesLoop: for (final M move : moves) {
            final B newBoard = move.apply(board);
            boardsGenerated.getAndIncrement();
            final GameState<M, B> newGameState = new GameState<>(move, newBoard);
            final LinkedNode<GameState<M, B>> newNode = new LinkedNode<>(newGameState, node);
            for (final MoveFilter<M, B> filter : gameFilters) {
                if (filter.shouldFilter(newNode)) {
                    final Object statisticsKey = filter.getStatisticsKey();
                    movesPrunedTotal.getAndIncrement();
                    movesPruned.get(statisticsKey).getAndIncrement();
                    continue movesLoop;
                }
            }
            
            if (game.isWin(newNode)) {
                addWin(newNode);
            } else {
                queueNode(newNode);
            }
        }
    }

}

File: csharp/Solitaire/Game/Search/GameTreeSearcher.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace Silnith.Game.Search
{
    /// <summary>
    /// Abstract base class for game tree search algorithms.
    /// This provides a common implementation of the logic to expand a single node
    /// in the game tree.
    /// </summary>
    /// <typeparam name="M">The move type for the game.</typeparam>
    /// <typeparam name="B">The board type for the game.</typeparam>
    public abstract class GameTreeSearcher<M, B> where M : IMove<B>
    {
        private readonly IGame<M, B> game;
        private readonly IEnumerable<IMoveFilter<M, B>> gameFilters;
        private long gameStatesExamined;
        private long boardsGenerated;
        private long movesPrunedTotal;
        private readonly ConcurrentDictionary<object, long> movesPruned;

        /// <summary>
        /// Initializes a game tree search run for the provided game.
        /// </summary>
        /// <param name="game">The game.</param>
        protected GameTreeSearcher(IGame<M, B> game)
        {
            this.game = game ?? throw new ArgumentNullException(nameof(game));
            gameFilters = this.game.GetFilters();
            gameStatesExamined = 0;
            boardsGenerated = 0;
            movesPrunedTotal = 0;
            movesPruned = new ConcurrentDictionary<object, long>();
            foreach (IMoveFilter<M, B> filter in gameFilters)
            {
                movesPruned[filter.StatisticsKey] = 0;
            }
        }

        /// <summary>
        /// Searches the game tree.
        /// </summary>
        /// <returns>All winning game states found.</returns>
        public abstract IEnumerable<IReadOnlyList<GameState<M, B>>> Search();

        /// <summary>
        /// Searches the game tree.
        /// </summary>
        /// <param name="cancellationToken">A cancellation token.</param>
        /// <returns>All winning game states found.</returns>
        public abstract Task<IEnumerable<IReadOnlyList<GameState<M, B>>>> SearchAsync(CancellationToken cancellationToken = default);

        /// <summary>
        /// The current size of the queue of nodes to be searched.
        /// </summary>
        protected abstract int QueueSize
        {
            get;
        }

        /// <summary>
        /// The current count of winning game states found.
        /// </summary>
        protected abstract int WinCount
        {
            get;
        }

        /// <summary>
        /// Prints statistics on the game tree search to the console.
        /// </summary>
        public void PrintStatistics()
        {
            Console.WriteLine(
                "Nodes examined: {0:N0}\n"
                + "Boards generated: {1:N0}\n"
                + "Moves pruned: {2:N0}\n"
                + "Queue size: {3:N0}\n"
                + "Wins: {4:N0}",
                gameStatesExamined,
                boardsGenerated,
                movesPrunedTotal,
                QueueSize,
                WinCount);
            foreach (IMoveFilter<M, B> filter in gameFilters)
            {
                object statisticsKey = filter.StatisticsKey;
                Console.WriteLine(
                    "Moves pruned by filter {0}: {1:N0}",
                    statisticsKey,
                    movesPruned[statisticsKey]);
            }
        }

        /// <summary>
        /// The total number of game states examined.
        /// </summary>
        public long NumberOfGameStatesExamined
        {
            get
            {
                return gameStatesExamined;
            }
        }

        /// <summary>
        /// The total number of boards generated.
        /// </summary>
        public long BoardsGenerated
        {
            get
            {
                return boardsGenerated;
            }
        }

        /// <summary>
        /// Queues a game tree node for searching.
        /// </summary>
        /// <param name="node">The game tree node to search.</param>
        protected abstract void QueueNode(LinkedNode<GameState<M, B>> node);

        /// <summary>
        /// Captures a sequence of game states that culminates in a win.
        /// </summary>
        /// <param name="node">The game state history that results in a win.</param>
        protected abstract void AddWin(IReadOnlyList<GameState<M, B>> node);

        /// <summary>
        /// Examines a single node in the game tree.  This enumerates all possible
        /// moves, applies the move filters, and examines unfiltered moves to see if
        /// they are a winning game state.  Winning game states are passed to
        /// <see cref="AddWin(IReadOnlyList{GameState{M, B}})"/>.  All other unfiltered
        /// game states are passed to
        /// <see cref="QueueNode(LinkedNode{GameState{M, B}})"/>.
        /// </summary>
        /// <param name="node">The node containing the game state to examine.</param>
        protected void ExamineNode(LinkedNode<GameState<M, B>> node)
        {
            Interlocked.Increment(ref gameStatesExamined);
            GameState<M, B> gameState = node.Value;
            B board = gameState.Board;
            IEnumerable<M> moves = game.FindAllMoves(node);
            foreach (M move in moves)
            {
                B newBoard = move.Apply(board);
                Interlocked.Increment(ref boardsGenerated);
                GameState<M, B> newGameState = new GameState<M, B>(move, newBoard);
                LinkedNode<GameState<M, B>> newNode = new LinkedNode<GameState<M, B>>(newGameState, node);
                bool broken = false;
                foreach (IMoveFilter<M, B> filter in gameFilters)
                {
                    if (filter.ShouldFilter(newNode))
                    {
                        Interlocked.Increment(ref movesPrunedTotal);
                        object key = filter.StatisticsKey;
                        long value = movesPruned[key];
                        while (!movesPruned.TryUpdate(key, value + 1, value))
                        {
                            value = movesPruned[key];
                        }
                        broken = true;
                        break;
                    }
                }
                if (broken)
                {
                    continue;
                }

                if (game.IsWin(newNode))
                {
                    AddWin(newNode);
                }
                else
                {
                    QueueNode(newNode);
                }
            }
        }

    }
}
